src/wp-includes/global-styles-and-settings.php

```diff
@@ -306,8 +306,30 @@ function wp_add_global_styles_for_blocks() {
 
 	$tree        = WP_Theme_JSON_Resolver::get_merged_data();
 	$block_nodes = $tree->get_styles_block_nodes();
+
+	$can_use_cached = ! wp_is_development_mode( 'theme' );
+	if ( $can_use_cached ) {
+		$settings_hash = _wp_get_global_settings_hash();
+		$cache_key     = 'wp_styles_for_blocks';
+		$cached        = get_site_transient( $cache_key );
+		if ( ! is_array( $cached ) ) {
+			$cached = array();
+		}
+	}
+
 	foreach ( $block_nodes as $metadata ) {
-		$block_css = $tree->get_styles_for_block( $metadata );
+		if ( $can_use_cached ) {
+			$metadata_hash   = md5( wp_json_encode( $metadata ) );
+			$style_cache_key = $settings_hash . '_' . $metadata_hash;
+			if ( isset( $cached[ $style_cache_key ] ) ) {
+				$block_css = $cached[ $style_cache_key ];
+			} else {
+				$block_css                  = $tree->get_styles_for_block( $metadata );
+				$cached[ $style_cache_key ] = $block_css;
+			}
```

Comment:
If the settings hash is part of the transient key, it doesn't need to be used for each block node key:

```suggestion
			if ( isset( $cached[ $metadata_hash ] ) ) {
				$block_css = $cached[ $metadata_hash ];
			} else {
				$block_css                  = $tree->get_styles_for_block( $metadata );
				$cached[ $metadata_hash ] = $block_css;
			}
```